### DISCUSSION TITLE
How to apply conditional sprinkles across multiple variants, states and themes?

### DISCUSSION THREAD

**Discussion by thedanchez** [vanilla-extract-css/vanilla-extract#1012] [Q&A]
I just recently discovered `vanilla-extract` and I really like what I have seen thus far! I'm in the process of creating a component library and am working on the theme for library. For context, the theme I'm trying to implement has two axes to it:

```typescript
// your typical light vs dark mode
export type ColorMode = "light" | "dark";

// the density value affects spacing, font-sizes, line-height, etc. across all custom components
export type Density = "narrow" | "regular" | "wide"; 
```

With the above, I'm looking to do something like the following for a custom `button` component with multiple variants:

```tsx
// imagine a button with multiple variants: primary | secondary
// and also include the aforementioned theme dimensions: colorMode and density
// I want to do styling for all the possible button states like the following:

styleVariants({
  borderRadius: 4, // applies to all variants
  primary: {
    color: { light: "...", dark: "..." },
    padding: { narrow: ..., regular: ..., wide: ... },
    ":hover": {
      color: { light: "...", dark: "..." },
      padding: { narrow: ..., regular: ..., wide: ... }
    },
    ":disabled": {
      color: { light: "...", dark: "..." },
      padding: { narrow: ..., regular: ..., wide: ... }
    }
  },
  secondary: {
    color: { light: "...", dark: "..." },
    padding: { narrow: ..., regular: ..., wide: ... },
    ":hover": {
      color: { light: "...", dark: "..." },
      padding: { narrow: ..., regular: ..., wide: ... }
    },
    ":disabled": {
      color: { light: "...", dark: "..." },
      padding: { narrow: ..., regular: ..., wide: ... }
    }
  },
})
```

I'm pretty sure I have to make use of `sprinkles` in some way, shape or form since it supports the conditional styling. I just don't know how to achieve the complex conditional styling for both `colorMode` and `density` in **all** button states like `default`, `hover` and `disabled`. Does anyone know how I can achieve this? Is this even possible currently?

**Comment by thedanchez** [comment]
@askoufis do you know if the above is possible using `@vanilla-extract/sprinkles` ?

**Reply by thedanchez** [reply]
@askoufis  if this is something that is possible, it would extremely useful to document what it looks like to execute it because this is a pretty complex use-case I think. Essentially, what I'm looking for here is a blend of the `styleVariants` and `sprinkles` APIs. If it's not possible, I feel like this would be an awesome feature to support if it doesn't exist because it provides users a very succinct way to express theme-driven values across the different states of a given element.

**Reply by thedanchez** [reply]
@mattcompiles @markdalgleish do you know if the above is possible with the current set of APIs in vanilla-extract?

**Reply by mattcompiles** [reply]
It looks like you want something like the [Recipe API](https://vanilla-extract.style/documentation/packages/recipes/).

**Reply by thedanchez** [reply]
@mattcompiles thanks for letting me know about Recipes! Don't know how I overlooked that completely. I'll give it a shot.

Side question, if it does end up working out, would I be able to export style tokens if I use the recipes API? What I would want to get as an outcome for tokens is, if an application developer is using my component library with these exported tokens, when they do `tokens.button.primary.color` or `tokens.button.primary.padding` they would get the appropriate underlying value depending on the active `colorMode` or `density` value

**Reply by thedanchez** [reply]
@mattcompiles so the Recipes API gets me super close to what I'm looking for. If I'm able to marry it with the Sprinkles API then I think my problem is solved and according to the docs this should be possible! However, whenever I try to do so, I'm seeing unexpected results. For example, I'm trying to do the following:

```tsx
const color = { ... }
const spacing = { ... }

export const lightMode = "light"
export const darkMode = "dark"

const colorModeProperties = defineProperties({
  conditions: {
    light: {},
    dark: { selector: `.${darkMode} &` },
  },
  defaultCondition: "light",
  properties: {
    backgroundColor: color,
    borderColor: color,
    borderLeftColor: color,
    borderRightColor: color,
    borderTopColor: color,
    borderBottomColor: color,
    caretColor: color,
    outlineColor: color,
    textDecorationColor: color,
  }
})

const densityProperties = defineProperties({
  conditions: {
    narrow: { selector: ".narrow &" },
    regular: {},
    wide: { selector: ".wide &" }
  },
  defaultCondition: "regular",
  properties: {
    height: spacing,
    padding: spacing,
  }
})

export const sprinkles = createSprinkles(colorModeProperties, densityProperties);

// Below wherever I am using Sprinkles in the Recipe is the key to achieving the outcome I'm looking for!
// But I am not getting the expected output!
export const buttonRecipe = recipe({
  base: {
    borderRadius: 4,
    padding: sprinkles({ padding: { narrow: 4, regular: 8, wide: 12 }})
  },
  variants: {
    variant: {
      primary: {
        backgroundColor: sprinkles({ backgroundColor: { light "blue5", dark: "blue1" } }),
        borderColor: color.blue6,
        boxShadow: "inset 0px -1px 0px rgba(0, 0, 0, 0.40), inset 0px 1px 0px rgba(255, 255, 255, 0.40)",
        color: color.white,
        ":hover": {
          // Being able to leverage sprinkles here is key as well!
          backgroundColor: sprinkles({ backgroundColor: { light "blue4", dark: "blue6" } }),,
          borderColor: color.blue5,
        },
        ":active": {
          backgroundColor: color.blue6,
          borderColor: color.blue7,
        },
        ":focus": {
          backgroundColor: color.blue5,
          borderColor: color.blue2,
        },
        ":disabled": {
          backgroundColor: color.blue5,
          borderColor: color.blue6,
        },
      }
    },
  },
  defaultVariants: {
    variant: "primary",
  }
});
```

What ends up being output in the CSS value for `padding` for example is the following: `button__1gypylr2l8 button_padding_4_narrow__1gypylr2k8 button_padding_8_regular__1gypylr2kl button_padding_12_wide__1gypylr2ky` which is clearly an invalid padding value as opposed to one of the expected `4`, `8`, or `12` values depending on the active `density` value.

Any ideas on how I'm getting that value? I believe I'm following the example in the docs that showcases the use of Recipe + Sprinkles correctly.

**Reply by mattcompiles** [reply]
The `sprinkles` function is returning a list of classes. In your example you're passing that to an individual property which doesn't make sense. If you want to combine existing classes (i.e. sprinkles) with new ones then you need to pass an array. I think you want something like the following:

```js
export const buttonRecipe = recipe({
  base: [
    {
      borderRadius: 4,
    },
    sprinkles({ padding: { narrow: 4, regular: 8, wide: 12 }})
  ],
  variants: {
    variant: {
      primary: [
        sprinkles({ backgroundColor: { light "blue5", dark: "blue1" } }),
        {
          borderColor: color.blue6,
          boxShadow: "inset 0px -1px 0px rgba(0, 0, 0, 0.40), inset 0px 1px 0px rgba(255, 255, 255, 0.40)",
          color: color.white,
        }
     ]
  },
  defaultVariants: {
    variant: "primary",
  }
});


```

**Reply by thedanchez** [reply]
Yes that worked! So the last bit is doing the same thing for the pseudo-class states: `:hover`, `:focus`, etc.

Is there a way to use Sprinkles for those states?

**Reply by mattcompiles** [reply]
You'd have to duplicate all your conditions e.g. `darkHover` and `lightHover`. You can do it, but it'd get pretty messy. 

Have you considered using a CSS variable theme to control your colors rather than conditions. e.g. `colors.primary` which changes depending on your light/dark theme setting?

**Reply by thedanchez** [reply]
@mattcompiles haven't considered that approach actually. Not sure I have a clear idea of it though exactly.

**Reply by mattcompiles** [reply]
@aquarazorda's example below shows something like what I mean. This is a very rough example. I'd recommend reading through all the theming docs to get an idea of what's possible. 

```js
export const [lightThemeClass, vars] = createTheme({
  color: {
    primary: 'blue',
    primaryHover: 'darkblue'
  },
})

export const darkThemeClass = createTheme(vars, {
  color: {
    primary: 'red',
    primaryHover: 'darkred'
  },
})

const colorModeProperties = defineProperties({
  properties: {
    backgroundColor: vars.color,  
  }
})

// ...

export const buttonRecipe = recipe({
  variants: {
    variant: {
      primary: [
        // Using your current primary background color from sprinkles
        sprinkles({ backgroundColor: primary }),
        {
          ':hover': {
             // Or reference from the theme vars directly
             backgroundColor: vars.primaryHover
          }
        }
     ]
  },
  defaultVariants: {
    variant: "primary",
  }
});
```

**Reply by thedanchez** [reply]
So if I needed to handle different `density` variations then that would entail having to create separate themes for those as well yes?

**Reply by mattcompiles** [reply]
Yeah you could model it that way. Themes are just groupings of CSS variables so having multiple isn't an issue.

**Reply by airtonix** [reply]
@mattcompiles in your example, what is the purpose of this: 

```ts
const colorModeProperties = defineProperties({
  properties: {
    backgroundColor: vars.color,  
  }
})
``` 

nvm. it's mentioned here: https://vanilla-extract.style/documentation/packages/sprinkles/

**Comment by aquarazorda** [comment]
I've finally figured it out, I think. I'm not sure if there's a better way or not, but I've managed to swap out Light/Dark themes with this one. https://vanilla-extract.style/documentation/packages/dynamic/

Basically, I've got this const

```typescript
export const themeVars = createThemeContract({
	colors: {
		hover: null,
		textPrimary: null,
		textSecondary: null,
		textTertiary: null,
		bg: null,
		bgSecondary: null,
		bgButton: null,
		borderMain: null,
	}
});
```

and then two consts, something like this - 

```typescript
const colors = {
	hover: '#555657',
	textPrimary: '#fff',
	textSecondary: '#787878',
	textTertiary: '#AAAAAA',
	bg: '#222323',
	bgSecondary: '#222323',
	bgButton: '#555657',
	borderMain: '#787878',
};

export const darkTheme = {
	colors
};
```

Then in root I'm changing it like this -

```typescript
export const [themeMode, setThemeMode] = createSignal('dark');

export default function Root() {
	let html;
	const value = createI18nContext(languageDict, 'en');

	onMount(() => {
		createEffect(() => setElementVars(html, themeVars, themeMode() === 'dark' ? darkTheme : lightTheme));
	});

	return (
		<Html lang="en" ref={html}>
```

`setElementVars` expects an object as a third argument, that's why I'm not using `createTheme` on dark and light themes, it returns string and an object with var names.
I'm not sure if there's a better way.
I hope it helps.

**Reply by thedanchez** [reply]
Thanks for taking the time to look into this @aquarazorda! I really appreciate the effort.

Yeah, I was wondering if there was a way to do this at compile time instead of at runtime with the use of `dynamic` is the thing. Considering the `sprinkles` allows for declaring conditions, I would think it would be possible to allow for nesting/cascading conditions but I haven't seen any examples of it so far which makes me worried it's not possible to do so (without going the runtime route I mean). From your example I would have to do the same for `density` as what you're doing for `colorMode` which would make things look pretty hairy I think.
